Cython/Utility/CppConvert.pyx

```diff
@@ -52,14 +52,24 @@ cdef vector[X] {{cname}}(object o) except *:
 #################### vector.to_py ####################
 
 cdef extern from *:
-    cdef cppclass vector "const std::vector" [T]:
+    cdef cppclass vector "std::vector" [T]:
         size_t size()
         T& operator[](size_t)
 
-@cname("{{cname}}")
-cdef object {{cname}}(vector[X]& v):
-    return [v[i] for i in range(v.size())]
+cdef extern from "Python.h":
+    void Py_INCREF(object)
+    list PyList_New(Py_ssize_t size)
+    void PyList_SET_ITEM(object list, Py_ssize_t i, object o)
 
+@cname("{{cname}}")
+cdef object {{cname}}(const vector[X]& v):
+    o = PyList_New(<Py_ssize_t>v.size())
+    cdef Py_ssize_t i
+    for i in range(v.size()):
+        item = v[i]
+        Py_INCREF(item)
+        PyList_SET_ITEM(o, i, item)
```

Comment:
I think we should add a size check here to make sure that the vector did not grow during iteration.

Reply:
Can this happen? Since this requires the GIL I do not see how a different thread would change the size + generally C++ standard containers are not thread safe.

Reply:
In principle the constructor/destructor of any of the C++ objects could resize the vector, or the the constructor of the `item` Python object could release the GIL (most likely by invoking the GC, and causing arbitrary Python code to get executed in a destructor somewhere).

Personally I'd take the view that pretty much all of C++ iteration relies on containers not managing to resize themselves during it, and that it isn't necessary to handle this corner case. But I'm pretty sure I could force it to happen if I tried.

Reply:
With that said, if we were going to test it I think the place to do it would be before indexing the vector - I don't think that the vector growing is a problem for the current code because the stop condition for the range is only evaluated once

Reply:
Right. This case is probably ok. If the vector grows, it will stop short, which is fine since we don't guarantee anything in that case except that it doesn't crash. If the vector shrinks, it would raise a C++ exception on indexing, which should then get turned into an `IndexError`. So that's fine, too.

Reply:
Yes probably you could force it. C++ containers are not thread safe, so users should manage this themselves. I do not see any way for us to make sure they are in this use case. Checking the size in front of each access would just make it harder to break it, but if you try hard you can still break it.

Reply:
> If the vector shrinks, it would raise a C++ exception on indexing, which should then get turned into an IndexError. So that's fine, too.
No it does not
v[i] just accesses out of bouns memory which is undefined behavior

v.at(i) raises an exception, but since vector is not thread safe it might resize in between the size check and the access in the at function

Also many of our iterations are based on iterators, which might get invalidated -> undefined behavior

Reply:
I think threading is not something we should care about. But modification during iteration is. As @da-woods pointed out, the vector might change while instantiating the Python item, which can execute arbitrary Python code. We should guard against that.

Reply:
We could use a `while` loop and add a second loop after it that uses `append()` if there is really anything left to append.

Reply:
I'm not sure a second loop is worthwhile. I'd argue for either:
* Use the code as-is (on the basis this anything that changes the vector during iteration is very unusual, and people who write such code should be ultra-careful about ownership and object lifetime)
* Swap `[i]` for `.at(I)`, on the basis that it should keep things fairly safe (i.e. raise an Python exception)

Personally I don't think it's sensible to try to handle too much beyond that. Any code that could change the size of vector could also go back and modify earlier elements, so we can never do more than just a "best-effort snapshot"

Reply:
I would argue that we should mimic the semantics of Python list iteration here. If you change a Python list while iterating over it, it will simply stop as soon as it has returned as many (or more) items as the current list length at this time. It may skip over items when you add or remove some, but it won't crash, and it will look at the list length on every iteration.

Note that this is not special-purpose code. It's generic list copying code that users won't even notice they are using most of the time. It should try to be as widely applicable as possible (but not a tad slower than strictly necessary). If we introduce potential bugs or non-obvious constraints here, then there is a high chance that it won't be noticed in user code until much, much later, when the right kind of data passes through it during the right phase of the moon.

Reply:
> Also many of our iterations are based on iterators, which might get invalidated -> undefined behavior

One final comment on this from me I think: for C++ linked lists, sets and maps, using iterators is the only way to traverse then and C++ iterators have a very strong requirement that the container won't be modified as you go. So for vector/string/deque we could definitely make it a bit more robust to changes, but for most other  types I don't think it's possible.

Reply:
@da-woods That's a convincing reason for those containers then. :)

Just keep in mind that C++ containers can easily be used by Cython developers as low-level data types even if their C++ knowledge is limited. If we can make it work for them by providing sensible semantics, we should.

Reply:
@da-woods What you wrote doesn't seem to apply to linked lists, though, e.g. https://www.cplusplus.com/reference/list/list/push_front/ – deques yes, maps and sets partly. Clearing always invalidates, for example, but inserting doesn't necessarily. So there are several modifications that users can apply to C++ containers without rendering their iterators invalid or useless.

Reply:
Yes linked lists are not that problematic. The iterator can simply check the next pointer each time. It is only problematic when the element the iterator currently points to is invalidated.